pkg/vere/io/ames.c

```diff
@@ -1173,6 +1189,193 @@ _fine_put_cache(u3_ames* sam_u, u3_noun pax, c3_w lop_w, u3_noun lis)
   }
 }
 
+static void
+_stun_stop(u3_ames* sam_u)
+{
+  switch ( sam_u->sun_u.sat_y ) {
+    case STUN_OFF: break;  //  ignore; already stopped
+    case STUN_TRYING: {
+      uv_timer_stop(&sam_u->sun_u.tim_u);
+    } break;
+    case STUN_KEEPALIVE: {
+      uv_timer_stop(&sam_u->sun_u.tim_u);
+    } break;
+    default: u3_assert(!"programmer error");
+  }
+  sam_u->sun_u.sat_y = STUN_OFF;
+}
+
+static void _stun_send_request(u3_ames*); // forward declaration
+
+static void
+_stun_timer_cb(uv_timer_t* tim_u)
+{
+  u3_ames* sam_u = (u3_ames*)(tim_u->data);
+  switch ( sam_u->sun_u.sat_y ) {
+    case STUN_OFF: {
+      //  ignore; stray timer (although this shouldn't happen)
+      u3l_log("stun: stray timer STUN_OFF");
+    } break;
+    case STUN_KEEPALIVE: {
+      sam_u->sun_u.sat_y = STUN_TRYING;
+      _stun_send_request(sam_u);
+    } break;
+    case STUN_TRYING: {
+      c3_d gap_d;
+      {
+        struct timeval tim_tv;
+        gettimeofday(&tim_tv, 0);
+        u3_noun now = u3_time_in_tv(&tim_tv);
+        u3_noun den = u3_time_in_tv(&sam_u->sun_u.sar_u);
+        gap_d = u3_time_gap_ms(den, now);
+      }
+      if ( gap_d >= (5 * 1000) ) {
+        _stun_stop(sam_u);
+        //  TODO inject event into arvo to ping repeatedly
+      }
+      else {
+        sam_u->sun_u.tim_u.data = sam_u;
+        uv_timer_start(&sam_u->sun_u.tim_u, _stun_timer_cb, 1*1000, 0);
+      }
+    } break;
+    default: u3_assert(!"programmer error");
+  }
+}
+
+typedef struct _u3_stun_send {
+  uv_udp_send_t req_u;  //  uv udp request handle
+  u3_ames* sam_u;       //  backpointer to driver state
+} u3_stun_send;
+
+static void
+_stun_send_cb(uv_udp_send_t *req_u, c3_i sas_i)
+{
+  u3_stun_send* snd_u = (u3_stun_send*)req_u;
+  if ( sas_i ) {
+    //  TODO take error handling action
+  }
+  else {
+    u3_ames* sam_u = snd_u->sam_u;
+    gettimeofday(&sam_u->sun_u.las_u, 0);  //  overwrite last sent date
+  }
+}
+
+static void
+_stun_on_response(u3_ames* sam_u) // TODO read arg
+{
+  u3_stun_state old_y = sam_u->sun_u.sat_y;
+  switch ( sam_u->sun_u.sat_y ) {
+    case STUN_OFF: break;       //  ignore; stray response
+    case STUN_KEEPALIVE: break; //  ignore; duplicate response
+    case STUN_TRYING: {
+      sam_u->sun_u.sat_y = STUN_KEEPALIVE;
+      sam_u->sun_u.tid_y[1]++;
+      uv_timer_stop(&sam_u->sun_u.tim_u);
+      sam_u->sun_u.tim_u.data = sam_u;
+      uv_timer_start(&sam_u->sun_u.tim_u, _stun_timer_cb, 25*1000, 0);
+    } break;
+    default: assert("programmer error");
+  }
+}
```

Comment:
This should either be called from `_ames_recv_cb()` (if we can always distinguish STUN responses preemptively), or from `_ames_hear()` (as a fallback if a packet is not a valid ames packet).

I believe the response will look like:

```
c3_y rep_y[20] = {
  0x1, 0x1,                // binding resp
  0x??, 0x??,              // len (including optional attributes)
  0x21, 0x12, 0xA4, 0x42,  // cookie
  ... original 12 bytes of randomness
  ... XOR-mapped address attribute
}
```

An ames header will never begin with `{ 0x1, 0x1 }` (binding success response) or `{ 0x1, 0x11 }` (binding error response), so that plus a sanity check of the constant "STUN cookie" should be enough to distinguish protocols.

Reply:
I'm going off the rfc (https://www.rfc-editor.org/rfc/rfc8489), and sanity checking that with packet captures (https://www.cloudshark.org/captures/4686b8430477).